providers/shared/fqdn.go

```diff
@@ -60,13 +66,13 @@ func fqdn(hostname string) (string, error) {
 		return strings.ToLower(strings.TrimSuffix(cname, ".")), nil
 	}
 
-	ips, err := net.LookupIP(hostname)
+	ips, err := net.DefaultResolver.LookupIP(ctx, "ip", hostname)
 	if err != nil {
 		errs = fmt.Errorf("%s: failed looking up IP: %w", errs, err)
 	}
 
 	for _, ip := range ips {
-		names, err := net.LookupAddr(ip.String())
+		names, err := net.DefaultResolver.LookupAddr(ctx, ip.String())
```

Comment:
The type of `ips` here changed from `[]net.IP` to `[]net.IPAddr` which doesn't have the same implementation of `String()`, since [net.IPAddr](https://pkg.go.dev/net#IPAddr) embeds `IP` and extends it with a zone.

If you want the exact same conversion as before you want to call `ip.IP.String()` I think.

```go

func (a *IPAddr) String() string {
	if a == nil {
		return "<nil>"
	}
	ip := ipEmptyString(a.IP)
	if a.Zone != "" {
		return ip + "%" + a.Zone
	}
	return ip
}
```

The original call to `net.LookupIP` stripped off the IPv6 zone: https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/lookup.go;l=195

```go
// LookupIP looks up host using the local resolver.
// It returns a slice of that host's IPv4 and IPv6 addresses.
func LookupIP(host string) ([]IP, error) {
	addrs, err := DefaultResolver.LookupIPAddr(context.Background(), host)
	if err != nil {
		return nil, err
	}
	ips := make([]IP, len(addrs))
	for i, ia := range addrs {
		ips[i] = ia.IP
	}
	return ips, nil
}
```

Reply:
> The type of `ips` here changed from `[]net.IP` to `[]net.IPAddr`...

Are you sure? Before, `ips` was initialized as:

```go
ips, err := net.LookupIP(hostname)
```

Now, it's initialized as:

```go
ips, err := net.DefaultResolver.LookupIP(ctx, "ip", hostname)
```

In both cases, the type of `ips` is the same: `[]IP`.
* `net.LookupIP`: https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/lookup.go;l=193-205
* `net.DefaultResolver.LookupIP`: https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/net/lookup.go;l=213-234

Reply:
Ah you are right for some reason I ended up focusing on LookupIPAddr as if it were called early, I think this was because `net.LookupIP` calls LookupIPAddr but `func (r *Resolver) LookupIP` does something else.

I don't have enough brainpower left for the day or network knowledge to know if that is significant, I was trying to quickly check if this worked the same as it did before.